lib/lte_link_control/lte_lc.c

```diff
@@ -768,6 +791,217 @@ static int parse_rrc_mode(const char *at_response,
+{
+	int err;
+	enum lte_lc_system_mode sys_mode;
+	u16_t multiplier = 0;
+
+	/* Lookup table to eDRX multiplier values, based on T_eDRX values found
+	 * in Table 10.5.5.32/3GPP TS 24.008. The actual value is
+	 * (multiplier * 10.24 s), except for the first entry which is handled
+	 * as a special case per note 3 in the specification.
+	 */
+	static const u16_t edrx_lookup_ltem[16] = {
+		0, 1, 2, 4, 6, 8, 10, 12, 14, 16, 32, 64, 128, 256, 256, 256
+	};
+	static const u16_t edrx_lookup_nbiot[16] = {
+		2, 2, 2, 4, 2, 8, 2, 2, 2, 16, 32, 64, 128, 256, 512, 1024
+	};
+
+	if (edrx_value == NULL) {
+		return -EINVAL;
+	}
+
+	err = lte_lc_system_mode_get(&sys_mode);
+	if (err) {
+		LOG_ERR("Failed to get system mode, error: %d", err);
+		return err;
+	}
+
+	switch (sys_mode) {
+	case LTE_LC_SYSTEM_MODE_LTEM: /* Fall through */
+	case LTE_LC_SYSTEM_MODE_LTEM_GPS:
+		multiplier = edrx_lookup_ltem[idx];
+		break;
+	case LTE_LC_SYSTEM_MODE_NBIOT: /* Fall through */
+	case LTE_LC_SYSTEM_MODE_NBIOT_GPS:
+		multiplier = edrx_lookup_nbiot[idx];
+		break;
+	case LTE_LC_SYSTEM_MODE_GPS: /* Fall through */
+	case LTE_LC_SYSTEM_MODE_NONE: /* Fall through */
+	default:
+		LOG_ERR("No LTE connection available in this system mode");
+		return -ENOTCONN;
+	}
+
+	*edrx_value = multiplier == 0 ? 5.12 : multiplier * 10.24;
+
+	return err;
+}
+
+/**@brief Parses an AT command response, and returns the current eDRX settings.
+ *
+ * @note It's assumed that the network only reports valid eDRX values when
+ *	 in each mode (LTE-M and NB1). There's no sanity-check of these values.
+ *
+ * @param at_response Pointer to buffer with AT response.
+ * @param cfg Pointer to where the eDRX configuration is stored.
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+static int parse_edrx(const char *at_response,
+		      struct lte_lc_edrx_cfg *cfg)
+{
+	int err;
+	u8_t idx;
+	struct at_param_list resp_list = {0};
+	char tmp_buf[5];
+	size_t len = sizeof(tmp_buf);
+	float ptw_multiplier;
+
+	if ((at_response == NULL) || (cfg == NULL)) {
+		return -EINVAL;
+	}
+
+	/* Confirm valid system mode and set Paging Time Window multiplier.
+	 * Multiplier is 1.28 s for LTE-M, and 2.56 s for NB-IoT, derived from
+	 * figure 10.5.5.32/3GPP TS 24.008.
+	 */
+	err = get_ptw_multiplier(&ptw_multiplier);
+	if (err) {
+		return err;
+	}
+
+	err = at_params_list_init(&resp_list, AT_CEDRXP_PARAMS_COUNT_MAX);
+	if (err) {
+		LOG_ERR("Could not init AT params list, error: %d", err);
+		return err;
+	}
+
+	/* Parse CEREG response and populate AT parameter list */
+	err = at_parser_max_params_from_str(at_response,
+					    NULL,
+					    &resp_list,
+					    AT_CEDRXP_PARAMS_COUNT_MAX);
+	if (err) {
+		LOG_ERR("Could not parse +CEDRXP response, error: %d", err);
+		goto clean_exit;
+	}
+
+	err = at_params_string_get(&resp_list, AT_CEDRXP_NW_EDRX_INDEX,
+				   tmp_buf, &len);
+	if (err) {
+		LOG_ERR("Failed to get eDRX configuration, error: %d", err);
+		return err;
```

Comment:
All exit points after `at_params_list_init` should rather call `goto clean_exit;` to avoid memory leak.

Reply:
Thanks, good catch